Microsoft.Azure.Cosmos/src/Resource/QueryResponses/CosmosQueryResponse.cs

```
@@ -3,10 +3,124 @@
     using System.Linq;
+    using Microsoft.Azure.Cosmos.Collections;
+    using Microsoft.Azure.Cosmos.Internal;
+
+    /// <summary>
+    /// The cosmos query response
+    /// </summary>
+    public class CosmosQueryResponse : IDisposable
+    {
+        private bool _isDisposed = false;
+        private INameValueCollection _responseHeaders = null;
+
+        /// <summary>
+        /// Create a <see cref="CosmosQueryResponse{T}"/>
+        /// </summary>
+        internal CosmosQueryResponse(
+            INameValueCollection responseHeaders,
+            Stream content,
+            string continuationToken)
+        {
+            this._responseHeaders = responseHeaders;
+            this.Content = content;
+            this.ContinuationToken = continuationToken;
+        }
+
+        internal CosmosQueryResponse(
+            Exception exception,
+            INameValueCollection responseHeaders = null)
+        {
+            this.ContinuationToken = null;
+            this.Content = null;
+            this._responseHeaders = responseHeaders;
+            this.Exception = exception;
+        }
+
+        /// <summary>
+        /// Gets the continuation token
+        /// </summary>
+        public virtual string ContinuationToken { get; protected set; }
+
+        /// <summary>
+        /// Contains the stream response of the operation
+        /// </summary>
+        public virtual Stream Content { get; protected set; }
+
+        /// <summary>
+        /// The exception if the operation failed.
+        /// </summary>
+        public virtual Exception Exception { get; }
+
+        /// <summary>
+        /// Gets the request charge for this request from the Azure Cosmos DB service.
+        /// </summary>
+        /// <value>
+        /// The request charge measured in request units.
+        /// </value>
+        public virtual double RequestCharge
+        {
+            get
+            {
+                if (this._responseHeaders == null)
```

I don't think this should happen and instead we should throw an exception.

---

For the exception code path this will not be set. I think it will be easier for users to do the same null check rather than deal with an exception. Also in the exceptionless code path we are trying to avoid throwing exceptions.